src/ir/return-utils.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ir_return_h
+#define wasm_ir_return_h
+
+#include "ir/utils.h"
+#include "wasm-builder.h"
+#include "wasm-traversal.h"
+#include "wasm.h"
+
+namespace wasm::ReturnUtils {
+
+// Removes values from both explicit returns and implicit ones (values that flow
+// from the body). This is useful after changing a function's type to no longer
+// return anything.
+struct ReturnValueRemover : public PostWalker<ReturnValueRemover> {
+  void visitReturn(Return* curr) {
+    auto* value = curr->value;
+    assert(value);
+    curr->value = nullptr;
+    Builder builder(*getModule());
+    replaceCurrent(builder.makeSequence(builder.makeDrop(value), curr));
+  }
+
+  void visitCall(Call* curr) {
+    handleReturnCall(curr, getModule()->getFunction(curr->target)->getSig());
+  }
+
+  void visitCallIndirect(CallIndirect* curr) {
+    handleReturnCall(curr, curr->heapType.getSignature());
+  }
+
+  void visitCallRef(CallRef* curr) {
+    Type targetType = curr->target->type;
+    if (!targetType.isSignature()) {
+      // We don't know what type the call should return, but it will also never
+      // be reached, so we don't need to do anything here.
+      return;
+    }
+    handleReturnCall(curr, targetType.getHeapType().getSignature());
+  }
+
+
+  template<typename T>
+  void handleReturnCall(T* curr, Signature sig) {
+    if (curr->isReturn) {
+      // This can no longer be a return call, as it calls something that returns
+      // a value, and we do not. Update the type (note we must handle the case
+      // of an unreachable child, and also this change may affect our parent, so
+      // refinalize the entire function).
+      curr->isReturn = false;
+      curr->type = sig.results;
+      curr->finalize();
+      refinalize = true;
+
+      replaceCurrent(Builder(*getModule()).makeDrop(curr));
```

Comment:
This needs to be followed by a `return` to avoid going to on execute subsequent code that should have been dead.

Reply:
Right, thanks, the semantics include the return :smile: ... Fixed.

Reply:
Hmm, and now that I think about it, this also needs to do the thing where it breaks out of the function body before doing the call to make sure exceptions propagate correctly.

Reply:
I don't understand why we shouldn't optimize return calls in monomorphized bodies. If you know the caller drops the result, you can change a return call in the callee to be a branch + call + drop sequence, which can allow further optimizations like monomorphizing the former return-callee now that we see that its result is also dropped.

Reply:
It can definitely help in some cases, yeah, but it can break others. Imagine that A calls B which calls A etc. in some very deep stack before it stops. If we monomorphize them both, then even if we save 90% of the work inside them, if the calls are no longer return calls then we'd be using potentially enough stack to error.

Reply:
Oh, I see. FWIW, inlining also has this problem because the frame sizes increase with the depth of inlining, even through tail calls. I don't think that's enough reason to avoid optimizing, though.

Reply:
I worry less about frame sizes increasing as that's generally going to be a "moderate" factor, like maybe the frame doubles in size. Maybe it even 10x grows somehow? But if we turn tail calls into normal calls for a language that does loops using tail calls then we can turn what were 1,000,000 loop iterations into 1,000,000 nested calls, can't we? :scream:

Reply:
Hmm, I guess if we wanted to avoid that problem, we would want to make the monomorphization of a function containing a return call contingent on the monomorphization of the return-callee as well. So the entire return-call graph reachable from the original monomorphized function would be monomorphized as a unit.

Reply:
Yeah, I think that's right.

Another option might be to only monomorphize in cases that we shrink the called function so much it will get inlined immediately, which eliminates a call.